### PR TITLE
test: Overhaul DSN tests

### PR DESCRIPTION
**Pull request by rhcarvalho**
- Use cmp.Diff to compare values and print easy to read diffs on error.
- Use table tests to consistently check the same things for multiple
inputs. Previously, some tests omitted parts that were covered by
others.
- Use consistent names and values matching DSN concepts. Helps with
reading and understanding tests.
- Use raw strings to avoid escaping quotes.
- Do not ignore errors from NewDsn.

### CODE REVIEW

File: dsn.go [hunk 1]
```diff
@@ -159,6 +159,7 @@ func (dsn Dsn) StoreAPIURL() *url.URL {
 
 // RequestHeaders returns all the necessary headers that have to be used in the transport.
 func (dsn Dsn) RequestHeaders() map[string]string {
+	// FIXME: why sentry_version=7?
```
Comment (rhcarvalho) [inline dsn.go]: @kamilogorek is this something like an API version? When does it change?

I'd like to replace the magic number literal with a named constant 😄

Comment (kamilogorek) [reply dsn.go]: Correct, it's minimal version of the API that's compatible with the SDK. in JS we call it `SENTRY_API_VERSION`

Comment (rhcarvalho) [reply dsn.go]: Thanks, I moved it into a constant.

Comment (rhcarvalho) [reply dsn.go]: #113

File: dsn_test.go [hunk 2]
```diff
@@ -3,36 +3,121 @@ package sentry
...
+		})
 	}
-	assertEqual(t, schemeHTTPS, dsn.scheme)
-	assertEqual(t, dsn.publicKey, "username")
-	assertEqual(t, dsn.secretKey, "password")
-	assertEqual(t, dsn.host, "domain")
-	assertEqual(t, dsn.path, "/foo/bar")
-	assertEqual(t, dsn.String(), url)
-	assertEqual(t, dsn.port, 8888)
-	assertEqual(t, dsn.projectID, 23)
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "AllFields" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR20

File: dsn_test.go [hunk 3]
```diff
@@ -3,36 +3,121 @@ package sentry
...
-}
-
-func TestDsnDefaultPort(t *testing.T) {
-	dsn, _ := NewDsn("https://u@d:1337/23")
-	assertEqual(t, dsn.port, 1337)
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "AllFields" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR20

File: dsn_test.go [hunk 4]
```diff
@@ -3,36 +3,121 @@ package sentry
...
-	dsn, _ = NewDsn("https://u@d/23")
-	assertEqual(t, dsn.port, 443)
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "MinimalSecure" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR33

File: dsn_test.go [hunk 5]
```diff
@@ -3,36 +3,121 @@ package sentry
...
-	dsn, _ = NewDsn("http://u@d/23")
-	assertEqual(t, dsn.port, 80)
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "MinimalInsecure" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR44

File: dsn_test.go [hunk 6]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
 	}
 }
 
-func TestValidDsnInsecure(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "MinimalInsecure" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR44

Except for the custom port, tested in "AllFields".

File: dsn_test.go [hunk 7]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	url := "http://username@domain:8888/42"
-	dsn, err := NewDsn(url)
-
-	if err != nil {
-		t.Error("expected dsn to be correctly created")
-	}
-	assertEqual(t, url, dsn.String())
-}
-
-func TestValidDsnNoPort(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "MinimalInsecure" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR44

File: dsn_test.go [hunk 8]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-
-	if err != nil {
-		t.Error("expected dsn to be correctly created")
-	}
-	assertEqual(t, 80, dsn.port)
-	assertEqual(t, url, dsn.String())
-	assertEqual(t, "http://domain/api/42/store/", dsn.StoreAPIURL().String())
-}
-
-func TestValidDsnPrefixed(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "AllFields" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR20

Except that "AllFields" uses HTTPS. The important part is that we now always check the `StoreAPIURL`.

File: dsn_test.go [hunk 9]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	url := "http://username@domain/prefixed/42"
-	dsn, err := NewDsn(url)
-
-	if err != nil {
-		t.Error("expected dsn to be correctly created")
-	}
-	assertEqual(t, "http://domain/prefixed/api/42/store/", dsn.StoreAPIURL().String())
-}
-
-func TestValidDsnInsecureNoPort(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "MinimalSecure" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR33

Note that the name of the test here was misleading. It said "Insecure" but the URL was "https".

File: dsn_test.go [hunk 10]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-
-	if err != nil {
-		t.Error("expected dsn to be correctly created")
-	}
-	assertEqual(t, 443, dsn.port)
-	assertEqual(t, url, dsn.String())
-	assertEqual(t, "https://domain/api/42/store/", dsn.StoreAPIURL().String())
-}
-
-func TestValidDsnNoPassword(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "MinimalSecure" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR33

This test and `TestValidDsnInsecureNoPort` only differed in the custom port. We test that we can do custom ports in "AllFields".

File: dsn_test.go [hunk 11]
```diff
@@ -3,36 +3,121 @@ package sentry
...
+}
+
+type invalidDsnTest struct {
+	in  string
+	err string // expected substring of the error
+}
+
+//nolint: gochecknoglobals
+var invalidDsnTests = map[string]invalidDsnTest{
+	"Empty":     {"", "invalid scheme"},
```
Comment (rhcarvalho) [inline dsn_test.go]: This was previously not covered.

The test documents the current behavior. Not sure if an empty string should be considered valid by `NewDsn`. @kamilogorek what do you think?

File: dsn_test.go [hunk 12]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-
 	if err != nil {
-		t.Error("expected dsn to be correctly created")
+		t.Fatal(err)
 	}
-	assertEqual(t, url, dsn.String())
-	assertEqual(t, "https://domain:8888/api/42/store/", dsn.StoreAPIURL().String())
-}
-
-func TestInvalidDsnInvalidUrl(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "BadURL" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR94

File: dsn_test.go [hunk 13]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	_, err := NewDsn("!@#$%^&*()")
-	_, ok := err.(*DsnParseError)
-
-	if ok != true {
-		t.Error("expected error to be of type DsnParseError")
-	}
-	assertStringContains(t, err.Error(), "invalid url")
-}
-
-func TestInvalidDsnInvalidScheme(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "BadScheme" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR95

File: dsn_test.go [hunk 14]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	_, err := NewDsn("ftp://username:password@domain:8888/1")
-	_, ok := err.(*DsnParseError)
-
-	if ok != true {
-		t.Error("expected error to be of type DsnParseError")
-	}
-	assertStringContains(t, err.Error(), "invalid scheme")
-}
-
-func TestInvalidDsnNoUsername(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "NoPublicKey" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR90

File: dsn_test.go [hunk 15]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	_, err := NewDsn("https://:password@domain:8888/23")
-	_, ok := err.(*DsnParseError)
-
-	if ok != true {
-		t.Error("expected error to be of type DsnParseError")
-	}
-	assertStringContains(t, err.Error(), "empty username")
-}
-
-func TestInvalidDsnNoHost(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "NoHost" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR91

File: dsn_test.go [hunk 16]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	_, err := NewDsn("https://username:password@:8888/42")
-	_, ok := err.(*DsnParseError)
-
-	if ok != true {
-		t.Error("expected error to be of type DsnParseError")
-	}
-	assertStringContains(t, err.Error(), "empty host")
-}
-
-func TestInvalidDsnInvalidPort(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "BadPort" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR97

File: dsn_test.go [hunk 17]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	_, err := NewDsn("https://username:password@domain:wat/42")
-	_, ok := err.(*DsnParseError)
-
-	if ok != true {
-		t.Error("expected error to be of type DsnParseError")
-	}
-	assertStringContains(t, err.Error(), "invalid port")
-}
-
-func TestInvalidDsnNoProjectId(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "NoProjectID2" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR93

File: dsn_test.go [hunk 18]
```diff
@@ -3,36 +3,121 @@ package sentry
...
+//nolint: gochecknoglobals
+var invalidDsnTests = map[string]invalidDsnTest{
+	"Empty":     {"", "invalid scheme"},
+	"NoScheme1": {"public:secret@:8888/42", "invalid scheme"},
+	// FIXME: NoScheme2's error message is inconsistent with NoScheme1; consider
+	// avoiding leaking errors from url.Parse.
+	"NoScheme2":     {"://public:secret@:8888/42", "missing protocol scheme"},
+	"NoPublicKey":   {"https://:secret@domain:8888/42", "empty username"},
+	"NoHost":        {"https://public:secret@:8888/42", "empty host"},
+	"NoProjectID1":  {"https://public:secret@domain:8888", "empty project id"},
```
Comment (rhcarvalho) [inline dsn_test.go]: This is new.

File: dsn_test.go [hunk 19]
```diff
@@ -3,36 +3,121 @@ package sentry
...
+	err string // expected substring of the error
+}
+
+//nolint: gochecknoglobals
+var invalidDsnTests = map[string]invalidDsnTest{
+	"Empty":     {"", "invalid scheme"},
+	"NoScheme1": {"public:secret@:8888/42", "invalid scheme"},
+	// FIXME: NoScheme2's error message is inconsistent with NoScheme1; consider
+	// avoiding leaking errors from url.Parse.
+	"NoScheme2":     {"://public:secret@:8888/42", "missing protocol scheme"},
```
Comment (rhcarvalho) [inline dsn_test.go]: This is new. Produces a different error than omitting the scheme altogether. Tied to `url.Parse` behavior.

File: dsn_test.go [hunk 20]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	_, err := NewDsn("https://username:password@domain:8888/")
-	_, ok := err.(*DsnParseError)
-
-	if ok != true {
-		t.Error("expected error to be of type DsnParseError")
-	}
-	assertStringContains(t, err.Error(), "empty project id")
-}
-
-func TestInvalidDsnInvalidProjectId(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Now covered by "BadProjectID" test case.
https://github.com/getsentry/sentry-go/pull/112/files#diff-eeb2e1d2b086ffdc5e8a6e60178702aaR96

File: dsn_test.go [hunk 21]
```diff
@@ -62,137 +147,15 @@ func TestDsnDeserializeInvalidJSON(t *testing.T) {
...
-	_, err := NewDsn("https://username:password@domain:8888/wbvdf7^W#$")
-	_, ok := err.(*DsnParseError)
-
-	if ok != true {
-		t.Error("expected error to be of type DsnParseError")
-	}
-	assertStringContains(t, err.Error(), "invalid project id")
-}
-
-func TestRequestHeadersWithoutPassword(t *testing.T) {
```
Comment (rhcarvalho) [inline dsn_test.go]: Renamed to `TestRequestHeadersWithoutSecretKey`.

File: dsn_test.go [hunk 22]
```diff
@@ -3,36 +3,121 @@ package sentry
...
+	"NoScheme2":     {"://public:secret@:8888/42", "missing protocol scheme"},
+	"NoPublicKey":   {"https://:secret@domain:8888/42", "empty username"},
+	"NoHost":        {"https://public:secret@:8888/42", "empty host"},
+	"NoProjectID1":  {"https://public:secret@domain:8888", "empty project id"},
+	"NoProjectID2":  {"https://public:secret@domain:8888/", "empty project id"},
+	"BadURL":        {"!@#$%^&*()", "invalid url"},
+	"BadScheme":     {"ftp://public:secret@domain:8888/1", "invalid scheme"},
+	"BadProjectID":  {"https://public:secret@domain:8888/wbvdf7^W#$", "invalid project id"},
+	"BadPort":       {"https://public:secret@domain:wat/42", "invalid port"},
+	"TrailingSlash": {"https://public:secret@domain:8888/42/", "invalid project id"},
```
Comment (rhcarvalho) [inline dsn_test.go]: This is new and documents the current behavior.

The Sentry UI displays the DSN without trailing slashes.

**Pull request merged**